### DISCUSSION TITLE
Application-managed configuration

### DISCUSSION THREAD

**Discussion by pfmoore** [twpayne/chezmoi#3661] [General]
I work mainly on Windows, and I'm considering moving to chezmoi to manage my config rather than my current "do it all manually" mess 😉 But my problem is that for a lot of the config I want to manage, I don't edit dotfiles directly. The applications I use include a "config manager" of some sort, which stores config in a file somewhere. So for me, the idea of editing the chezmoi-managed copy of my config and then applying that to my application doesn't really feel natural (and in cases where there is a GUI for changing settings, it's significantly less user friendly).

For example, Windows Terminal has a settings GUI, and stores its config in `%LOCALAPPDATA%\Packages\Microsoft.WindowsTerminal_8wekyb3d8bbwe\LocalState\settings.json`. The file is manually editable, but the GUI is *far* easier. How would I set things up so that I can manage Windows Terminal settings with chezmoi, while still using the GUI to make changes? I'm fine with a workflow that involves having to do some sort of "compare current settings with saved ones, then sync chezmoi with the live settings" process.

Other applications that do this include `git` (`git config`), `pip` (`pip config`), and VS Code. So it's not an uncommon model, and I'm sure people have solved it, but I couldn't find a good discussion in the docs (which is probably my fault for not knowing where to look).

I don't have any immediate need to integrate this requirement with the more complex features of chezmoi, like templates. Right now I don't even have a need for things like that, although I'm sure I'll find uses as I get more familiar with it 🙂 I'd understand if advanced features were incompatible with using app-provided settings management, although I'd like it if I could set things up so that I didn't wreck my setup if I forgot and (for example) accidentally used an app config command on something I was using templates for.

**Comment by Lockszmith-GH** [comment]
You don't NEED to edit it with chezmoi.

Once you add a file to chezmoi with `chezmoi add` it is in **managed** state.

`chezmoi status` will then let you know whethere managed states have changed, and you can either `re-add` (update the managed copy - equivalent to saving after `chezmoi edit`) or `apply` (which will revert any changes back to the `managed` state).

In the diagram [here](https://www.chezmoi.io/user-guide/command-overview/#daily-commands), replace `chezmoi edit` with:
* Modifty file (whatever method)
* use `chezmoi add <changed file>` or `chezmoi re-add` to apply changes in managed files into the managed state (in the diagram: working copy)

**Reply by pfmoore** [reply]
OK, thanks. I think I follow what you mean - I'll give it a try and see if my understanding is correct.

**Reply by Lockszmith-GH** [reply]
If you are comfortable with git, think of chezmoi's working directory as an additional layer between your actual user profile configurations and git. That's how I actually got into using chezmoi.
There is so much more than just this extra layer, but that's a good place to start.

**Reply by halostatue** [reply]
Also note `chezmoi merge <file>` or `chezmoi merge-all` (this is what I use most often).

**Comment by bradenhilton** [comment]
You may be interested in [Handle configuration files which are externally modified](https://www.chezmoi.io/user-guide/manage-different-types-of-file/#handle-configuration-files-which-are-externally-modified).

The example is for Linux but it also applies to Windows (with a few path tweaks).

The basic idea is to store in the config file in chezmoi's source state and symlink it to its expected location. When the program then modifies its own config it will actually modify the file stored in your dotfiles repo. I can only assume that you would also need sufficient privileges to create symlinks with your user account, either by enabling Developer Mode or manually granting your account the privilege.

I don't use this personally, despite primarily using Windows, as I don't mind editing the files manually, and I also use a template for my VS Code config.

**Reply by saharshbhansali** [reply]
I've got a little bit of added complexity in implementing this solution. 

I want to backup a (chromium-based) browser's configuration. There are extensions and other sensitive data in those files, so I wanted to encrypt the entire browser's profile. 

Would it be possible to both encrypt and implement the solution you've suggested? Is there any other way I can do this?

**Comment by aniketgm** [comment]
In addition to other's comments. I suggest better start looking others configs [here](https://github.com/topics/chezmoi?o=desc&s=updated) I too recently discovered chezmoi and I'm also on windows. I suggest spending some time through the docs as well. The docs is really good.
Also explore [this](https://www.chezmoi.io/links/articles/) and other links on that site.

From what I've learnt till now is that chezmoi has a different system of it's own to manage dotfiles. Like renaming files for e.g. `dot_bashrc`, etc. which should not be a concern as long we don't have to take care of how it renames it back to the bashrc file in home directory.
That is why I mentioned above that spend a bit of time reading the docs. Hope this helps. Peaceout ✌️
